### PR TITLE
Fix geometry and indexing of new/modified roadway nodes/links

### PR DESCRIPTION
**Pull request by e-lo**
This PR contains major refactoring of several key pieces of RoadwayNetwork to be more efficient, resilient and understandable.

Improves:
- Efficiency in apply managed lane, add/delete and modify property changes
- Decomposes many monolithic functions and files (still many more to address)
- Adds `num_managed_lane_links()` method to easily test for managed lanes (#294)
- DRY principles
- Efficiency in tests: uses small network when able to, uses fixtures for network so only need to read in once (#300)
- Comprehensiveness in tests: actual assert statements, more tests
- Organization of tests: into smaller files so that can easily run them and understand what each does
- Organization of model-network functions into separate file as a model for decomposing some of roadway network which is...massive
- added documentation at function level
- formatting

Fixes:
- How geometries are modified when nodes/links are added/deleted/modified and for managed lanes (#297)
- How indices are modified when nodes/links are added/deleted/modified and for managed lanes
- A LOT of smaller bugs

...still a lot to do and that this PR doesn't fix.

### CODE REVIEW

**Comment by e-lo** [comment]
@lmz @DavidOry @i-am-sijia  - This is passing the roadway tests on my machine. Def not perfect but better and more resilient.

**Comment by i-am-sijia** [comment]
tests\test_utils.py fails at [test_point_from_xy()](https://github.com/wsp-sag/network_wrangler/blob/54d7cb3c32f64ff27925c6aff21906f2a90bbbae/tests/test_utils.py#L105):

image

After I change the `output_crs` to 4269, it passes. I think this is related to Issue #302

File: tests/test_roadway/test_changes/test_roadway_feature_change.py [hunk 1]
```diff
@@ -0,0 +1,282 @@
...
+
+def test_add_adhoc_field(request, small_net):
+    """
+    Makes sure new fields can be added in the API and be saved and read in again.
+    """
+    print("\n--Starting:", request.node.name)
+    net = copy.deepcopy(small_net)
+    net.links_df["my_ad_hoc_field"] = 22.5
+
+    print("Network with field...\n ", net.links_df["my_ad_hoc_field"][0:5])
```
Comment (i-am-sijia) [inline tests/test_roadway/test_changes/test_roadway_feature_change.py:178]:
```suggestion
    print("Network with field...\n ", net.links_df["my_ad_hoc_field"].iloc[0:5])
```

File: tests/test_roadway/test_changes/test_roadway_feature_change.py [hunk 2]
```diff
@@ -0,0 +1,282 @@
...
+
+    assert net.links_df["my_ad_hoc_field"][0] == 22.5
```
Comment (i-am-sijia) [inline tests/test_roadway/test_changes/test_roadway_feature_change.py:180]:
```suggestion
    assert net.links_df["my_ad_hoc_field"].iloc[0] == 22.5
```

File: tests/test_roadway/test_changes/test_roadway_feature_change.py [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+    """
+
+    print("\n--Starting:", request.node.name)
+    net = copy.deepcopy(small_net)
+    ok_properties_change = [{"property": "lanes", "change": 1}]
+    bad_properties_change = [{"property": "my_random_var", "change": 1}]
+    bad_properties_existing = [{"property": "my_random_var", "existing": 1}]
+
+    with pytest.raises(ValueError):
+        net.validate_properties(bad_properties_change)
```
Comment (i-am-sijia) [inline tests/test_roadway/test_changes/test_roadway_feature_change.py:274]:
```suggestion
        net.validate_properties(net.links_df, bad_properties_change)
```

File: tests/test_roadway/test_changes/test_roadway_feature_change.py [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+
+    with pytest.raises(ValueError):
+        net.validate_properties(ok_properties_change, require_existing_for_change=True)
```
Comment (i-am-sijia) [inline tests/test_roadway/test_changes/test_roadway_feature_change.py:277]:
```suggestion
        net.validate_properties(net.links_df, ok_properties_change, require_existing_for_change=True)
```

File: tests/test_roadway/test_changes/test_roadway_feature_change.py [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+
+    with pytest.raises(ValueError):
+        net.validate_properties(bad_properties_existing, ignore_existing=False)
```
Comment (i-am-sijia) [inline tests/test_roadway/test_changes/test_roadway_feature_change.py:280]:
```suggestion
        net.validate_properties(net.links_df, bad_properties_existing, ignore_existing=False)
```

File: network_wrangler/roadwaynetwork.py [hunk 6]
```diff
@@ -1284,183 +1572,505 @@ def validate_properties(
         Returns: boolean value as to whether the properties dictonary is valid.
         """
 
-        validation_error_message = []
-
+        valid = True
+        validation_error_message = ""
```
Comment (i-am-sijia) [inline network_wrangler/roadwaynetwork.py:1576]: changing this to string does not work with line 1596 below - `validation_error_message.append()`, in addition, this variable is not being used in any logging.

File: network_wrangler/utils/__init__.py [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+    "offset_location_reference",
+    "haversine_distance",
+    "create_unique_shape_id",
+    "location_reference_from_nodes",
+    "line_string_from_location_references",
+    "net_to_mapbox",
+    "get_point_geometry_from_linestring",
+    "point_df_to_geojson",
+    "point_from_xy",
+    "topological_sort" "update_points_in_linestring",
```
Comment (i-am-sijia) [inline network_wrangler/utils/__init__.py:29]:
```suggestion
    "topological_sort",
    "update_points_in_linestring",
```

File: network_wrangler/transitnetwork.py [hunk 8]
```diff
@@ -630,17 +634,16 @@ def _apply_individual_change(project_dictionary: dict):
...
-    ) -> Union(None, TransitNetwork):
+    def apply_python_calculation(self, pycode: str) -> "TransitNetwork":
         """
         Changes roadway network object by executing pycode.
 
         Args:
             pycode: python code which changes values in the roadway network object
-            in_place: update self or return a new roadway network object
         """
+        net = copy.deepcopy(self)
```
Comment (i-am-sijia) [inline network_wrangler/transitnetwork.py:644]: Why do we need a deepcopy here? The same function in roadwaynetwork.py does not deepcopy.

File: network_wrangler/transitnetwork.py [hunk 9]
```diff
@@ -1060,29 +1069,30 @@ def _apply_transit_feature_change_frequencies(
...
-        self._apply_transit_feature_change_routing(
+        TransitNetwork._apply_transit_feature_change_routing(
+            net,
             trip_ids=trip_ids,
             properties={
                 "existing": node_ids,
                 "set": RoadwayNetwork.get_managed_lane_node_ids(node_ids),
             },
-            in_place=in_place,
         )
```
Comment (i-am-sijia) [inline network_wrangler/transitnetwork.py:1097]: no `return net`?

File: network_wrangler/transitnetwork.py [hunk 10]
```diff
@@ -1025,22 +1039,17 @@ def _apply_transit_feature_change_routing(
...
-        else:
-            updated_network = copy.deepcopy(self)
-            updated_network.feed.shapes = shapes
-            updated_network.feed.stops = stops
-            updated_network.feed.stop_times = stop_times
-            return updated_network
+            net.feed.shapes = shapes
+            net.feed.stops = stops
+            net.feed.stop_times = stop_times
+            return net
```
Comment (i-am-sijia) [inline network_wrangler/transitnetwork.py:1045]: The indentation of this return seems off, should we move it outside of the if?

File: examples/single/link.json [hunk 11]
```diff
@@ -15,7 +15,7 @@
     "roadway": "residential",
     "length": 80.935,
     "maxspeed": "",
-    "name": "Arkwright Street,North Rivoli Street",
+    "name": "Arkwright Street",
```
Comment (i-am-sijia) [inline examples/single/link.json:18]: Why shorten the name?

**Review by i-am-sijia — CHANGES_REQUESTED** [review]
- ran all tests locally, proposed changes necessary for tests to pass
- the returns in transitnetwork.py seem inconsistent, after introducing `net` which is a deepcopy of `self`
- the roadwaynetwork.py has tons of changes, which I have not yet detailly reviewed, but the tests all passed.
- the `SCRATCH_DIR` in "test_transit.py" and "test_scenario.py" does not exist by default. It will be created after test_io.py is run which calls the pytest fixture. We need to be careful about the test dependencies when migrating to GitHub Actions.

File: examples/single/link.json (hunk 11 above)
Comment (e-lo) [reply examples/single/link.json:18]: Just to differentiate from the larger network when I'm reading errors (this net is really not "real")

File: network_wrangler/transitnetwork.py (hunk 10 above)
Comment (e-lo) [reply network_wrangler/transitnetwork.py:1045]: good catch

File: tests/test_performance.py [hunk 12]
```diff
@@ -0,0 +1,42 @@
...
+
+EX_DIR = r"/Users/elizabeth/Downloads/standard_roadway_pre_base_project_cards"
+EX_DIR = r"/Users/elizabeth/Documents/urbanlabs/MetCouncil/working/network_wrangler/examples/stpaul"
+
+LINKS = os.path.join(EX_DIR, "link.json")
+NODES = os.path.join(EX_DIR, "node.geojson")
+SHAPES = os.path.join(EX_DIR, "shape.geojson")
+
+
+def read_ex_net(benchmark):
```
[the remark this answers is not in the document]
Comment (e-lo) [reply tests/test_performance.py:16]: Benchmark is a keyword for pytest benchmark for timing the tests.

File: network_wrangler/transitnetwork.py [hunk 13]
```diff
@@ -1060,29 +1069,30 @@ def _apply_transit_feature_change_frequencies(
...
-        q = self.feed.frequencies.trip_id.isin(freq["trip_id"])
-        if in_place:
-            self.feed.frequencies.loc[q, properties["property"]] = build_value
-        else:
-            updated_network = copy.deepcopy(self)
-            updated_network.loc[q, properties["property"]] = build_value
-            return updated_network
+        q = net.feed.frequencies.trip_id.isin(freq["trip_id"])
+
+        net.loc[q, properties["property"]] = build_value
```
[the remark this answers is not in the document]
Comment (e-lo) [reply network_wrangler/transitnetwork.py:1074]: good catch!

File: tests/test_roadway/test_changes/test_pycode.py [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+    WranglerLogger.debug(f"Before Change:\n{_link_sel[_show_fields]}")
+
+    net = net.apply(
+        {
+            "category": "Calculated Roadway",
+            "project": "megaroads",
+            "pycode": _pycode,
+        }
+    )
+    WranglerLogger.debug(f"After Change:\n{_link_sel[_show_fields]}")
```
[the remark this answers is not in the document]
Comment (e-lo) [reply tests/test_roadway/test_changes/test_pycode.py:54]: Fixed using slightly different approach which looks up the inx

**Comment by e-lo** [comment]
@i-am-sijia I think this is ready for you to re-review

File: network_wrangler/transitnetwork.py [hunk 15]
```diff
@@ -1060,30 +1070,30 @@ def _apply_transit_feature_change_frequencies(
...
-        self, trip_ids: pd.Series, node_ids: list, in_place: bool = True
-    ) -> Union(None, TransitNetwork):
+        self,
+        trip_ids: pd.Series,
+        node_ids: list,
+    ) -> TransitNetwork:
+
         # Traversed nodes without a stop should be negative integers
-        all_stops = self.feed.stops[TransitNetwork.STOPS_FOREIGN_KEY].tolist()
+        all_stops = net.feed.stops[TransitNetwork.STOPS_FOREIGN_KEY].tolist()
```
Comment (i-am-sijia) [inline network_wrangler/transitnetwork.py:1085]:
```suggestion
        net = copy.deepcopy(self)
        all_stops = net.feed.stops[TransitNetwork.STOPS_FOREIGN_KEY].tolist()
```

**Review by i-am-sijia — CHANGES_REQUESTED** [review]
Just one small issue. Otherwise looks good.

**Review by i-am-sijia — APPROVED** [review]

**Comment by e-lo** [comment]
📓 that most of the transit tests aren't passing yet and will be worked on in a separate PR

**Comment by i-am-sijia** [comment]
@e-lo I added a lot of transit related codes to MTC's network wrangler: https://github.com/BayAreaMetro/network_wrangler/tree/generic_agency/network_wrangler, so that it can actually apply real transit project cards. it might be useful to include those in the transit related PR.

**Comment by e-lo** [comment]
@i-am-sijia I created an issue where you can add more detail if you think it would be useful: https://github.com/wsp-sag/network_wrangler/issues/319

**Pull request merged**